geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/ReplicateRegionFunction.java

```diff
@@ -0,0 +1,362 @@
+
+  static class ThreadSleeper implements Serializable {
+    void millis(long millis) throws InterruptedException {
+      Thread.sleep(millis);
+    }
+  }
+
+  @VisibleForTesting
+  public void setClock(Clock clock) {
+    this.clock = clock;
+  }
+
+  @VisibleForTesting
+  public void setThreadSleeper(ThreadSleeper ts) {
+    this.threadSleeper = ts;
+  }
+
+  @Override
+  public String getId() {
+    return ID;
+  }
+
+  @Override
+  public boolean hasResult() {
+    return true;
+  }
+
+  @Override
+  public boolean isHA() {
+    return false;
+  }
+
+  @Override
+  public CliFunctionResult executeFunction(FunctionContext<String[]> context) {
+    final Object[] args = context.getArguments();
+    if (args.length < 5) {
+      throw new IllegalStateException(
+          "Arguments length does not match required length.");
+    }
+    final String regionName = (String) args[0];
+    final String senderId = (String) args[1];
+    final boolean isCancel = (Boolean) args[2];
+    long maxRate = (Long) args[3];
+    int batchSize = (Integer) args[4];
+
+    final InternalCache cache = (InternalCache) context.getCache();
+    final Region region = cache.getRegion(regionName);
+    if (isCancel) {
+      return cancelReplicateRegion(context, region, senderId);
+    }
+
+    GatewaySender sender = cache.getGatewaySender(senderId);
+    if (!sender.isRunning()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Sender not running");
+    }
+
+    if (!sender.isParallel() && !((InternalGatewaySender) sender).isPrimary()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+          "Sender is serial and not primary. 0 entries replicated.");
+    }
+
+    try {
+      return executeReplicateFunctionInNewThread(context, region, sender, maxRate, batchSize);
+    } catch (InterruptedException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution canceled.");
+    } catch (ExecutionException e) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          "Execution failed: " + e.getMessage());
+    }
+  }
+
+  private CliFunctionResult executeReplicateFunctionInNewThread(FunctionContext<String[]> context,
+      Region region, GatewaySender sender, long maxRate, int batchSize)
+      throws InterruptedException, ExecutionException {
+    Callable<CliFunctionResult> callable =
+        new ReplicateRegionCallable(context, region, sender, maxRate, batchSize);
+    FutureTask<CliFunctionResult> futureTask = new FutureTask<>(callable);
+    Thread t = new Thread(futureTask);
+    String origThreadName = Thread.currentThread().getName();
+    try {
+      t.setName(getReplicateRegionFunctionThreadName(region.getName(),
+          sender.getId()));
+    } finally {
+      Thread.currentThread().setName(origThreadName);
```

Comment:
I'm not sure I understand why you're getting the calling thread name and re-setting it in the finally block. Its not being changed in the try anywhere.

Reply:
My bad. It makes no sense.